Replacing emulation with cross-compilation in the build-arch-emu.yaml workflow to speed up SuiteSparse builds

MigunovaAnastasia1 opened the pull request:
# SuiteSparse cross-compilation

## Supported architectures

All architectures for which SuiteSparse cross-dependencies exist in the Ubuntu repository are supported. Namely: 
* amd64
* armhf
* ppc64el
* s390x
* riscv64

## Results of replacing emulation with cross-compilation

Replacing emulation with cross-compilation gave a significant speedup in building SuiteSparse libraries for different architectures. The build is so fast that it is now possible to build absolutely all SuiteSparse libraries, including GraphBLAS and LAGraph, on different architectures. Already including GraphBLAS and LAGraph, it now takes only about 30 minutes to run build-arch-emu.yaml (now called 'cross-compile.yaml') workflow.

@DrTimothyAldenDavis also, the ability to build and test GraphBLAS and LAGraph may help you to solve the  RISCV testing problem for the following PR DrTimothyAldenDavis/GraphBLAS#381.

## Identified problem in LAGraph 

### Problem
By running the cross-compile.yaml workflow it was found that testing of LAGraph on **armhf** and **riscv64** architectures works correctly, but on the rest architectures **ppc64el** and **s390x** fails (see the result of the cross-compile.yaml [run](https://github.com/MigunovaAnastasia1/SuiteSparse-cross-compile/actions/runs/15065884178)).

### Localisation of the problem 
The identified problem was localised by a binary search (using the `git bisect` command) and found in the commit 1fd4ef8.

mmuetzel commented:
Nice! 👍

Maybe, it would make sense to keep one or two of the non-emulated Alpine Linux runners to keep testing against MUSL (compared to glibc on Ubuntu)...

vkutuev commented:
> Maybe, it would make sense to keep one or two of the non-emulated Alpine Linux runners to keep testing against MUSL (compared to glibc on Ubuntu)...

What is the best option for integrating the build and testing on Alpine Linux into the same workflow file or creating a separate one?

mmuetzel commented:
It might be best to keep the existing Alpine Linux file (and strip the emulated runners).
Then, add a new .yaml file with the rules for the cross-builds on Ubuntu.

MigunovaAnastasia1 committed:
removed unnecessary architectures

Signed-off-by: Migunova Anastasia <anastasiamigunova76@gmail.com>

MigunovaAnastasia1 committed:
added cross-compilation

Signed-off-by: Migunova Anastasia <anastasiamigunova76@gmail.com>

MigunovaAnastasia1 committed:
fixed matrix in build-arch-emu

Signed-off-by: Migunova Anastasia <anastasiamigunova76@gmail.com>

MigunovaAnastasia1 commented:
> It might be best to keep the existing Alpine Linux file (and strip the emulated runners). Then, add a new .yaml file with the rules for the cross-builds on Ubuntu.

Done.
I kept armv7, x86 and loongarch64 architectures in your build-arch-emu workflow. The last one is kept, as my cross-compilation based workflow doesn't  support it.

mmuetzel commented:
Thank you. That looks cleaner to me now.

Just asking: Why is there a configuration for `amd64` in the build matrix? Is it different from the existing "native" CI runners on Ubuntu?

@DrTimothyAldenDavis: Could you please approve running the workflows of this PR?

MigunovaAnastasia1 committed:
removed amd64 from build matrix

Signed-off-by: Migunova Anastasia <anastasiamigunova76@gmail.com>

MigunovaAnastasia1 commented:
> Just asking: Why is there a configuration for `amd64` in the build matrix? Is it different from the existing "native" CI runners on Ubuntu?

Apologies, just forgot to remove amd64 from the build matrix and yes, there are no differences from existing ‘native’ CI runners on Ubuntu.

DrTimothyAldenDavis commented:
Thanks for the updates.  

can you sign the SuiteSparse Contributor Agreement and send it to me at drtimothyaldendavis@gmail.com?

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/dev/Contributor_License/SuiteSparse%20Individual%20Contributor%20License%20Agreement%20(20241011).pdf

MigunovaAnastasia1 commented:
> can you sign the SuiteSparse Contributor Agreement and send it to me at [drtimothyaldendavis@gmail.com](mailto:drtimothyaldendavis@gmail.com)?

I apologise for forgetting to send the agreement. I did.

mmuetzel commented on .github/workflows/cross-compile.yaml:
```diff
@@ -0,0 +1,310 @@
+jobs:
+
+  cross-compile:
+    runs-on: ubuntu-latest
```
Since you are adding the Ubuntu repositories for Noble Nombat, would it make sense to pin this to `runs-on: ubuntu-24.04`?

DrTimothyAldenDavis commented:
> The identified problem was localised by a binary search (using the git bisect command) and found in the commit https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/1fd4ef8c95b83862c10c2e04df4b58cfbfccc3dc. 


Correction:  it's a gcc bug.   No need to split the yaml file.  See the details in my comments below.

mmuetzel commented on .github/workflows/cross-compile.yaml:
```diff
@@ -0,0 +1,310 @@
+        include:
+          - arch: armhf
+            target_system: arm-linux-gnueabihf
+            cross-toolchain: g++-9-arm-linux-gnueabihf
+            fortran-cross-toolchain: gfortran-10-arm-linux-gnueabihf
+            cc: "arm-linux-gnueabihf-gcc-9"
+            cxx: "arm-linux-gnueabihf-g++-9"
+            fc: "arm-linux-gnueabihf-gfortran-10"
```
Just out of interest: Is there a reason why you pin the compiler to these (old) versions here?
And similarly for the other host targets.

DrTimothyAldenDavis commented:
I suspect the bug is occurring because the ppc architecture triggers a bug in gcc. I just found this a week ago and haven't updated Suitesparse with this change:

https://github.com/DrTimothyAldenDavis/GraphBLAS/commit/15a92919662539c89a35291aaca7f95059ab3ab6

The problem is a failed OpenMP atomic in the symbolic phase of mxm. 

So I changed the GraphBLAS cmake so that it will refuse to compile on the Power processor with gcc.   Everything works fine with clang

DrTimothyAldenDavis commented:
See also https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/92bae92bea60c8ccaae4a8c0c939135f205e2036/Source/mxm/factory/GB_AxB_saxpy3_symbolic_template.c#L221-L233 

I've just copied over the latest dev2 branch of GraphBLAS into SuiteSparse.  This will cause the CI to fail early:  in the cmake when GraphBLAS is compiled.  The cmake script will report a failure if gcc is used to compile GraphBLAS on the Power processor.

Can the CI be changed to use clang or xlc instead of gcc, for the cross compilation?

DrTimothyAldenDavis commented on .github/workflows/cross-compile.yaml:
```diff
@@ -0,0 +1,310 @@
+            ccache-max: 64M
+          - arch: ppc64el
+            target_system: powerpc64le-linux-gnu
+            cross-toolchain: g++-9-powerpc64le-linux-gnu 
```
gcc has a bug that causes GraphBLAS to fail on the Power processor.  Is there a clang toolchain that can be used instead?

DrTimothyAldenDavis commented on .github/workflows/cross-compile.yaml:
```diff
@@ -0,0 +1,310 @@
+            ccache-max: 64M
+          - arch: s390x
+            target_system: s390x-linux-gnu
+            cross-toolchain: g++-9-s390x-linux-gnu 
```
I haven't tried gcc on the s390x with GraphBLAS, but I'm guessing it has the same bug as it does on the Power CPU.  Can you try a clang toolchain instead?

mmuetzel replied:
Untested: It might be possible to install the `clang` package (additionally to the GCC cross toolchain) and set `CC="clang -target ${{ matrix.target_system }}"` and `CXX="clang++ -target ${{ matrix.target_system }}"`.
The easiest way to do that is probably by changing the lines that start with `cc:` and `cxx:` in the `include` section of the `matrix` definition.

DrTimothyAldenDavis commented:
I have a Power8-based server, which is how I found this gcc bug.  I hacked my cmake script in the GraphBLAS/dev2 branch to enable gcc, and it exactly replicates the errors seen here in LAGraph.  The errors are the same on the s390x so I'm guessing it's the same gcc bug.  So replacing gcc with clang for both the powerpc and s390x builds will likely fix the issue (or provide a workaround for the gcc bug, to be more precise).

It's not common to find a bug in a compiler but I'm fairly certain that's what's happening here.

mmuetzel commented on .github/workflows/cross-compile.yaml:
```diff
@@ -0,0 +1,310 @@
+        include:
+          - arch: armhf
+            target_system: arm-linux-gnueabihf
+            cross-toolchain: g++-9-arm-linux-gnueabihf
+            fortran-cross-toolchain: gfortran-10-arm-linux-gnueabihf
```
Afaict, the value of `fortran-cross-toolchain` is only used where the value of `cross-toolchain` is also used.
Would it make sense to simplify the matrix slightly by combining the two?
``` suggestion
            cross-toolchain: g++-9-arm-linux-gnueabihf gfortran-10-arm-linux-gnueabihf
```

And similarly for the other host targets.

mmuetzel commented on .github/workflows/cross-compile.yaml:
```diff
@@ -0,0 +1,310 @@
+        arch: [armhf, ppc64el, s390x, riscv64]
+        include:
+          - arch: armhf
+            target_system: arm-linux-gnueabihf
```
Maybe, rename `target_system` to `target-triple`. Imho, that would make it easier to understand what it is without having to look at its value.

MigunovaAnastasia1 committed:
resolved the issue of using ubuntu-latest
made it possible to not replace ubuntu-latest in the ‘runs-on’ field with a specific version of ubuntu

Signed-off-by: Migunova Anastasia <anastasiamigunova76@gmail.com>

DrTimothyAldenDavis commented:
Oops!  I meant to authorize the workflow run but I clicked on "merge" instead.  I'd like to unmerge it, not because I don't want to accept it (I do), but because it's still in progress and I'd rather wait until all the tests pass.  Sorry for any confusion.  I'll try reverting the merge ... hopefully I won't break anything the process ...

DrTimothyAldenDavis commented:
Ack ... I thought that "revert" would place this pull request back into the "open" state, but it didn't.  So I broke it after all. 
 Sorry!

Can you reopen the pull request again?

MigunovaAnastasia1 commented:
> Ack ... I thought that "revert" would place this pull request back into the "open" state, but it didn't. So I broke it after all. Sorry!
> 
> Can you reopen the pull request again?

Sure, no problem!
